File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireHookCommandRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires hook files to declare and run harness validation commands.
 */
public class RequireHookCommandRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireHookCommand";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        JsonNode allowedCmds = catNode.get("parameters").get("allowedCommands").get("maven");
        String expectedCmd = allowedCmds == null || allowedCmds.size() == 0 ? "" : allowedCmds.get(0).asText();
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        String prePushPath = catNode.get("parameters").get("prePushHook").asText();
        Path prePush = root.resolve(prePushPath);
        if (!HarnessCheckHelper.isSafeRegularFile(root, prePush)) {
            return List.of();
        }
        String prePushText = HarnessCheckHelper.readFile(root, prePush);
        String declaredCmd = extractHookCommand(prePushText);
        List<Finding> prePushIssues = declaredCmd.isEmpty()
                ? List.of(new Finding(severity, CATEGORY, "pre-push hook must declare Harness validation command"))
                : !expectedCmd.equals(declaredCmd)
                ? List.of(new Finding(severity, CATEGORY, "pre-push hook declares unsupported validation command: " + declaredCmd))
                : !prePushText.contains(declaredCmd)
                ? List.of(new Finding(severity, CATEGORY, "pre-push hook must run the declared validation command"))
                : List.of();
        String preCommitPath = catNode.get("parameters").get("preCommitHook").asText();
        Path preCommit = root.resolve(preCommitPath);
        if (!HarnessCheckHelper.isSafeRegularFile(root, preCommit)) {
            return prePushIssues;
        }
        String preCommitText = HarnessCheckHelper.readFile(root, preCommit);
        Pattern fullStackPattern = Pattern.compile("(^|\\s)(uv|bun|gradle|mvn)(\\s|$)|\\.\\./gradlew|harnessValidate|harness_validate\\.py|harness-validate\\.ts");
        List<Finding> preCommitIssues = fullStackPattern.matcher(preCommitText).find()
                ? List.of(new Finding(severity, CATEGORY, "pre-commit hook must not run full stack validation commands"))
                : List.of();
        return Stream.concat(prePushIssues.stream(), preCommitIssues.stream()).toList();
    }

    private String extractHookCommand(String text) {
        for (String line : text.split("\\R")) {
            if (line.startsWith("# Harness validation command: ")) {
                return line.substring("# Harness validation command: ".length()).trim();
            }
        }
        return "";
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireHookCommandRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText

/**
 * Rule that requires hooks to declare and run validation commands.
 */
class RequireHookCommandRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireHookCommand"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> = buildSet {
		val category = "requireHookCommand"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return@buildSet
		val parametersObj = catObj["parameters"]?.jsonObject ?: return@buildSet
		val messagesObj = catObj["messages"]?.jsonObject ?: return@buildSet
		val allowedCmdObj = parametersObj["allowedCommands"]?.jsonObject
		val allowedCmds = HarnessCheck.Companion.stringArrayFrom(allowedCmdObj, "gradle")
		val allowedPreCommitCmdObj = parametersObj["allowedPreCommitCommands"]?.jsonObject
		val allowedPreCommitCmds = HarnessCheck.Companion.stringArrayFrom(allowedPreCommitCmdObj, "gradle")
		val prePushPath = HarnessCheck.Companion.stringFrom(parametersObj, "prePushHook")
		val preCommitPath = HarnessCheck.Companion.stringFrom(parametersObj, "preCommitHook")

		return buildSet<Finding> {
			val prePushHook = root / prePushPath
			if (prePushHook.isRegularFile()) {
				val text = prePushHook.readText()
				val command = text.lineSequence().firstOrNull { it.startsWith("# Harness validation command: ") }?.removePrefix("# Harness validation command: ")?.trim() ?: ""
				when {
					command.isEmpty() -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "missingDeclaration").takeIf { it.isNotEmpty() } ?: "pre-push hook must declare Harness validation command"
						add(Finding(severity, category, msg))
					}
					command !in allowedCmds -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "unsupportedCommand").takeIf { it.isNotEmpty() } ?: "pre-push hook declares unsupported validation command: $command"
						add(Finding(severity, category, msg))
					}
					!text.contains(command) -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "commandNotRun").takeIf { it.isNotEmpty() } ?: "pre-push hook must run the declared validation command"
						add(Finding(severity, category, msg))
					}
				}
			}

			val preCommitHook = root / preCommitPath
			if (preCommitHook.isRegularFile()) {
				val text = preCommitHook.readText()
				val command = text.lineSequence().firstOrNull { it.startsWith("# Harness validation command: ") }?.removePrefix("# Harness validation command: ")?.trim() ?: ""
				when {
					command.isEmpty() && allowedPreCommitCmds.isNotEmpty() -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "missingDeclaration").takeIf { it.isNotEmpty() } ?: "pre-commit hook must declare validation command"
						add(Finding(severity, category, msg))
					}
					command.isNotEmpty() && command !in allowedPreCommitCmds -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "unsupportedCommand").takeIf { it.isNotEmpty() } ?: "pre-commit hook declares unsupported validation command: $command"
						add(Finding(severity, category, msg))
					}
					command.isNotEmpty() && !text.contains(command) -> {
						val msg = HarnessCheck.Companion.stringFrom(messagesObj, "commandNotRun").takeIf { it.isNotEmpty() } ?: "pre-commit hook must run the declared validation command"
						add(Finding(severity, category, msg))
					}
				}
			}
		}.toList()
	}
}
